app/models/app_version.rb

```diff
@@ -0,0 +1,48 @@
+class AppVersion
+  include Comparable
+
+  DECIMAL_INTEGER = /(?:0|[1-9]\d*)/
+  VERSION_REGEX = /\A(?:#{DECIMAL_INTEGER}\.){0,2}#{DECIMAL_INTEGER}\z/ # iOS rules currently, but can be relaxed from here
+  MAX_LENGTH = 18 # iOS rules
+
+  attr_reader :version_number
+
+  def self.from_a(ary)
+    new(ary.map(&:to_s).join("."))
+  end
+
+  def self.from_pg_array(value)
+    from_a(value.scan(/\d+/))
+  end
+
+  def initialize(version_number)
+    raise "version_number must be a string, integer, or AppVersion" unless [String, Integer, AppVersion].any? do |t|
+      version_number.is_a?(t)
+    end
+    version_number = version_number.to_s
+    raise "version_number must be present" if version_number.blank?
+    raise "version_number is too long" if version_number.length > MAX_LENGTH
+    raise "version_number does not conform to format" unless version_number.match(VERSION_REGEX)
```

Comment:
So if someone changes/tightens any of these constraints after the fact, we'll `rescue` below in `FeatureCompletion` and get `nil` on the version. What are the downstream effects of that, and were does/should the nil-handling occur? (I'm having trouble tracing all of the entry points -- like is `define_build` actually used in a non-spec context?)

Reply:
I'm mostly worried about a code bug / developer error resulting in something being shown that shouldn't because the version-handling breaks down somewhere.

Reply:
So we've got a not-null constraint and presence validation on the FeatureCompletions table which will make sure that feature completions will always persist with well-formed versions. define_build will not rescue the exception because it doesn't run through the attributes API, so you'll get a hard foreground error. If I need to elevate that to validations in order to put better client errors in place, I can, but I suspect I can require the clients to validate the version format such that 500 errors are acceptable if a client pitches us a bad version.

Reply:
And yeah, the entrypoints will be twofold:

* Migrations from client side will create/mutate feature_completions
* Mobile clients will make visitor requests with an app_name, version_number and built_at in the payload which will be used to `define_build` and then the resulting app_build will be passed into the `for_presentation` scope for display (plus whatever equivalent thing needs doing to modify the split weighting registry to respect feature incompletions)

Reply:
I think it's the latter that might break when you go from `request` -> `version` (nil) -> `to_pg_array` (in `for_presentation`) -- again only if the data at rest used to be valid and no longer upon decoding / casting it to an `AppVersion`.

Reply:
It's not a huge risk, nor a direct result of anything in the current code -- but things do drift over time in surprising ways, and there is nothing stopping someone from coming in and "correcting" these `raise` calls in a way that invalidates existing production data.

Reply:
Oh wait, ignore me. I see now that it's `define_build` that will blow up with `AppVersion.new` -- already-persisted feature completions won't break that comparison because we're _not_ loading them up again at that point and just reading straight from the table. 🤦‍♂️ 

I got it, feel free to ignore me. I'll wrap my head around all of this eventually. 😅

Reply:
Deserializing probably would blow up in surprising ways, but you’d have already fixed my tests to make a change to a more restrictive format. I based my format on an Apple technical publication, and to my eye there’s no air gap between what Apple will allow and what we will allow in either direction, so I’d be surprised to find this getting more restrictive, if anything less to support 3-dot versions and alpha/beta/rc. Doing the textual stuff would require a migration to allow the db column to store text so that’d be the opportunity to true up platter representation with new format.